SU2_CFD/src/fluid/CFluidFlamelet.cpp

```diff
@@ -0,0 +1,245 @@
+  for (size_t i_aux = 0; i_aux < n_user_scalars; i_aux++) {
+    table_scalar_names[n_control_vars + i_aux] = config->GetUserScalarName(i_aux);
+  }
+
+  config->SetLUTScalarNames(table_scalar_names);
+
+  /*--- we currently only need 1 source term from the LUT for the progress variable
+        and each auxiliary equations needs 2 source terms ---*/
+  n_table_sources = 1 + 2 * n_user_scalars;
+
+  table_source_names.resize(n_table_sources);
+  table_sources.resize(n_table_sources);
+  table_source_names[I_SRC_TOT_PROGVAR] = "ProdRateTot_PV";
+  /*--- No source term for enthalpy ---*/
+
+  /*--- For the auxiliary equations, we use a positive (production) and a negative (consumption) term:
+        S_tot = S_PROD + S_CONS * Y ---*/
+
+  for (size_t i_aux = 0; i_aux < n_user_scalars; i_aux++) {
+    /*--- Order of the source terms: S_prod_1, S_cons_1, S_prod_2, S_cons_2, ...---*/
+    table_source_names[1 + 2 * i_aux] = config->GetUserSourceName(2 * i_aux);
+    table_source_names[1 + 2 * i_aux + 1] = config->GetUserSourceName(2 * i_aux + 1);
+  }
+
+  config->SetLUTSourceNames(table_source_names);
+
+  look_up_table = new CLookUpTable(config->GetFileNameLUT(), table_scalar_names[I_PROGVAR], table_scalar_names[I_ENTH]);
+
+  n_lookups = config->GetNLookups();
+  table_lookup_names.resize(n_lookups);
+  for (int i_lookup = 0; i_lookup < n_lookups; ++i_lookup) {
+    table_lookup_names[i_lookup] = config->GetLUTLookupName(i_lookup);
+  }
+
+  source_scalar.resize(n_scalars);
+  lookup_scalar.resize(n_lookups);
+
+  Pressure = value_pressure_operating;
+
+  PreprocessLookUp();
+}
+
+CFluidFlamelet::~CFluidFlamelet() {
+  delete look_up_table;
+}
+
+/*--- do a lookup for the list of variables in table_lookup_names, for visualization purposes ---*/
+unsigned long CFluidFlamelet::SetScalarLookups(const su2double* val_scalars) {
+  su2double enth = val_scalars[I_ENTH];
+  su2double prog = val_scalars[I_PROGVAR];
+
+  string name_enth = table_scalar_names[I_ENTH];
+  string name_prog = table_scalar_names[I_PROGVAR];
+
+  /*--- perform table lookup ---*/
+  unsigned long exit_code = look_up_table->LookUp_XY(table_lookup_names, lookup_scalar, prog, enth);
+
+  return exit_code;
+}
+
+/*--- set the source terms for the transport equations ---*/
+unsigned long CFluidFlamelet::SetScalarSources(const su2double* val_scalars) {
+  table_sources[0] = 0.0;
+
+  string name_enth = table_scalar_names[I_ENTH];
+  string name_prog = table_scalar_names[I_PROGVAR];
+
+  /*--- value for the progress variable and enthalpy ---*/
+  su2double enth = val_scalars[I_ENTH];
+  su2double prog = val_scalars[I_PROGVAR];
+
+  /*--- perform table lookup ---*/
+  unsigned long exit_code = look_up_table->LookUp_XY(varnames_Sources, val_vars_Sources, prog, enth);
+
+  /*--- the source term for the progress variable is always positive by construction, but we clip from below  --- */
+  source_scalar[I_PROGVAR] = max(EPS, table_sources[I_SRC_TOT_PROGVAR]);
+  source_scalar[I_ENTH] = 0.0;
```

Comment:
Handling source terms in the fluid model class also sounds like a bad idea.
Is this chained in any way to the computation of primitives? Please move it to the preprocessing or even better in the philosophical sense move this to the source numerics which currently does very little.
You can either pass the fluid model to the numerics or just the table, it also looks like this has a very simple Jacobian.

Reply:
OK, I moved this to preprocessing so the lookups for variables and source terms are together.  I removed all source term stuff from cfluidflamelet.